src/lmic/lmic.c

```diff
@@ -723,7 +723,7 @@ static CONST_TABLE(u1_t, macCmdSize)[] = {
 static u1_t getMacCmdSize(u1_t macCmd) {
     if (macCmd < 2)
         return 0;
-    if ((macCmd - 2) >= LENOF_TABLE(macCmdSize))
+    if ((macCmd - 2) >= (u1_t)LENOF_TABLE(macCmdSize))
```

Comment:
Sorry, can't accept this. I deliberately want to compare to the length of the table. Casting to (u1_t) breaks this if the table accidentally grows. Again, the calculation (since this is C) must always be done at full width, so there's no point in doing this unless the (u1_t) changes the answer; which it had better not do.

Reply:
Thanks, change is now `if (((size_t)(macCmd - 2)) >= LENOF_TABLE(macCmdSize))`
because `LENOF_TABLE()` is `sizeof()/sizeof()` *and* `macCmd` is always `>=2`.

Reply:
Thanks @d-a-v.  Looks like the compiler is really trying to make sure that our code will work on 37-bit ones-complement machines. Sigh.

Type casts are always dangerous.  Can you show the warning here? I think that `(macCmd - 2u)` would be enough to shut up the compiler. I speculate that they're promoting `macCmd` to int (rather than unsigned) before doing the subtract. Using `2u` would force a promote to unsigned on both sides.  (The code as written is, in fact, correct, because of the check at line 724, but GCC has forgotten how to confirm that, and is worried that `macCmd-2` might be negative. )

But to save time for you, may I suggest a transformation that ought to be warningless, and perhaps more efficient?

```c++
if (macCmd >= 2) {
  const unsigned macCmdMinus2 = macCmd - 2u;
  if (macCmdMinus2 < LENOF_TABLE(macCmdSize)) {
    // macCmd in table, fetch it's size.
    return TABLE_GET_U1(macCmdSize, macCmdMinus2);
} else {
  // macCmd too small or too large: return zero. Zero is
  // never a legal command size, so it signals an error to the
  // caller.
  return 0;
}
```

The promotion of `macCmd` in this case will be to `unsigned`, followed by an unsigned subtract which we store without truncation into `macCmdMinus2`. Indexing by `unsigned` is generally as or more efficient than indexing by `u1_t` (because the index is already the right width).

(This kind of headache is why I tend to avoid narrower-than-default-width types for parameters and computations; the promotion rules tend to drive pedantic compilers crazy -- wanting typecasts when I find typecasts to be more dangerous than helpful. But... I'm not going to make that kind of change in the LMIC, it is what it is.)

Reply:
Here is the warning:
```
.../arduino-lmic/src/lmic/lmic.c: In function 'getMacCmdSize':
.../arduino-lmic/src/lmic/lmic.c:726:22: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     if ((macCmd - 2) >= LENOF_TABLE(macCmdSize))
                      ^~
```

I updated the PR with your proposal above, slightly changed (so return is always "called")
```c++
static u1_t getMacCmdSize(u1_t macCmd) {
    if (macCmd >= 2) {
        const unsigned macCmdMinus2 = macCmd - 2u;
        if (macCmdMinus2 < LENOF_TABLE(macCmdSize)) {
            // macCmd in table, fetch it's size.
            return TABLE_GET_U1(macCmdSize, macCmdMinus2);
        }
    }
    // macCmd too small or too large: return zero. Zero is
    // never a legal command size, so it signals an error
    // to the caller.
    return 0;
}
```